addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/internal/discovery/AirVisualNodeDiscoveryService.java

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.openhab.binding.airvisualnode.internal.discovery;
+
+import java.net.UnknownHostException;
+import java.util.Collections;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryServiceCallback;
+import org.eclipse.smarthome.config.discovery.ExtendedDiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.airvisualnode.AirVisualNodeBindingConstants;
+import org.openhab.binding.airvisualnode.internal.config.AirVisualNodeConfig;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import jcifs.netbios.NbtAddress;
+import jcifs.smb.SmbFile;
+
+/**
+ * Autodiscovery for AirVisual Node by searching for a host advertised with the NetBIOS name 'AVISUAL-<SerialNumber>'.
+ *
+ * @author Victor Antonovich - Initial contribution
+ */
+@Component(service = DiscoveryService.class, immediate = true)
+public class AirVisualNodeDiscoveryService extends AbstractDiscoveryService implements ExtendedDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(AirVisualNodeDiscoveryService.class);
+
+    public static final String AVISUAL_WORKGROUP_NAME = "MSHOME";
+
+    private static final Pattern AVISUAL_NAME_PATTERN = Pattern.compile("^AVISUAL-([^/]+)$");
+
+    private ScheduledFuture<?> backgroundDiscoveryFuture;
+
+    private DiscoveryServiceCallback discoveryServiceCallback;
+
+    public AirVisualNodeDiscoveryService() {
+        super(Collections.singleton(AirVisualNodeBindingConstants.THING_TYPE_AVNODE), 600, true);
+    }
+
+    @Override
+    public void setDiscoveryServiceCallback(DiscoveryServiceCallback discoveryServiceCallback) {
+        this.discoveryServiceCallback = discoveryServiceCallback;
+    }
+
+    @Override
+    protected void startScan() {
+        logger.debug("Starting scan");
+        scheduler.execute(this::scan);
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting background discovery");
+        backgroundDiscoveryFuture = scheduler.scheduleWithFixedDelay(this::scan, 0, 300, TimeUnit.SECONDS);
```

Comment:
Are you sure that the library you are using is doing the scan in the same way as windows clients do it? Because if you do a periodic scan every 5 minutes of the whole network your machine could be subject to be blocked by a net intrusion system in certain network setups. So I am unsure whether we should keep the automatic scan without user interaction in here.

Reply:
Well, the discovery process just requests SMB servers list from the local master browser using JCIFS library. Not sure how this operation will be treated by a net intrusion systems if repeated though. Do you recommend to disable periodic autodiscovery here?

Reply:
I am no 100% sure. I just have in mind that we do not want discovery processes to "flood" the network without a user-interaction. A background scan works without a user interaction. If you would move it to `startScan` the user has to initiate the scan. But in your scenario is a corner case I guess and I am not a windows expert. Querying the master browser for the list of servers every 5min doesn't sound too bad though. Let's leave it in for now, I just wanted to understand what is actually happening here.